From 5b9d0e69b9c898126a6e0977759e77cd5f83a90e Mon Sep 17 00:00:00 2001
From: Vivek Kasireddy <vivek.kasireddy@intel.com>
Date: Wed, 9 Nov 2022 13:44:39 -0800
Subject: [PATCH 24/28] ui/gtk: Include the position info while setting the ui
 info

In situtations where the Guest uses multiple displays/outputs, this
position info is useful for aligning the Guest's outputs with that of
the Host's. This position info corresponds with the x and y fields
mentioned in the spec:
https://github.com/oasis-tcs/virtio-spec/blob/master/virtio-gpu.tex#L343

Cc: Dongwon Kim <dongwon.kim@intel.com>
Cc: Gerd Hoffmann <kraxel@redhat.com>
Signed-off-by: Vivek Kasireddy <vivek.kasireddy@intel.com>
---
 ui/gtk.c | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/ui/gtk.c b/ui/gtk.c
index 59295fb7f9..eb10f1954a 100644
--- a/ui/gtk.c
+++ b/ui/gtk.c
@@ -748,12 +748,31 @@ static gboolean gd_window_close(GtkWidget *widget, GdkEvent *event,
     return TRUE;
 }
 
+static void gd_set_window_position(VirtualConsole *vc, QemuUIInfo *info)
+{
+    GdkDisplay *dpy = gtk_widget_get_display(vc->gfx.drawing_area);
+    GdkWindow *window;
+    GdkMonitor *monitor;
+    GdkRectangle geometry;
+
+    if (!gtk_widget_get_realized(vc->gfx.drawing_area)) {
+        return;
+    }
+
+    window = gtk_widget_get_window(vc->gfx.drawing_area);
+    monitor = gdk_display_get_monitor_at_window(dpy, window);
+    gdk_monitor_get_geometry(monitor, &geometry);
+    info->xoff = geometry.x;
+    info->yoff = geometry.y;
+}
+
 static void gd_set_ui_refresh_rate(VirtualConsole *vc, int refresh_rate)
 {
     QemuUIInfo info;
 
     info = *dpy_get_ui_info(vc->gfx.dcl.con);
     info.refresh_rate = refresh_rate;
+    gd_set_window_position(vc, &info);
     dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
 }
 
@@ -764,6 +783,7 @@ static void gd_set_ui_size(VirtualConsole *vc, gint width, gint height)
     info = *dpy_get_ui_info(vc->gfx.dcl.con);
     info.width = width;
     info.height = height;
+    gd_set_window_position(vc, &info);
     dpy_set_ui_info(vc->gfx.dcl.con, &info, true);
 }
 
-- 
2.34.1

